tests/extra_source_and_tests/test_2d_pulsatile_poiseuille_flow/pulsatile_poiseuille_flow.cpp

```diff
@@ -0,0 +1,389 @@
+        std::vector<Vecd> inner_wall_shape;
+        inner_wall_shape.push_back(Vecd(-2.0*BW, 0.0));
+        inner_wall_shape.push_back(Vecd(-2.0*BW, DH));
+        inner_wall_shape.push_back(Vecd(DL + 2.0*BW, DH));
+        inner_wall_shape.push_back(Vecd(DL + 2.0*BW, 0.0));
+        inner_wall_shape.push_back(Vecd(-2.0*BW, 0.0));
+
+        multi_polygon_.addAPolygon(outer_wall_shape, ShapeBooleanOps::add);
+        multi_polygon_.addAPolygon(inner_wall_shape, ShapeBooleanOps::sub);
+    }
+};
+
+ /**
+ * @brief 	Main program starts here.
+ */
+int main(int ac, char *av[])
+{
+    /**
+     * @brief Build up -- a SPHSystem --
+     */
+    SPHSystem sph_system(system_domain_bounds, resolution_ref);
+    sph_system.setGenerateRegressionData(false);
+    sph_system.handleCommandlineOptions(ac, av)->setIOEnvironment();
+    /**
+     * @brief Material property, particles and body creation of fluid.
+     */
+    FluidBody water_block(sph_system, makeShared<WaterBlock>("WaterBody"));
+    water_block.defineParticlesAndMaterial<BaseParticles, WeaklyCompressibleFluid>(rho0_f, c_f, mu_f);
+    water_block.generateParticles<ParticleGeneratorLattice>();
+    /**
+     * @brief 	Particle and body creation of wall boundary.
+     */
+    SolidBody wall_boundary(sph_system, makeShared<WallBoundary>("Wall"));
+    wall_boundary.defineParticlesAndMaterial<SolidParticles, Solid>();
+    wall_boundary.generateParticles<ParticleGeneratorLattice>();
+
+    ObserverBody velocity_observer(sph_system, "VelocityObserver");
+    velocity_observer.generateParticles<ParticleGeneratorObserver>(observer_location);
+    /** topology */
+    InnerRelation water_block_inner(water_block);
+    ContactRelation water_block_contact(water_block, {&wall_boundary});
+    ContactRelation velocity_observer_contact(velocity_observer, {&water_block});
+    ComplexRelation water_block_complex(water_block_inner, water_block_contact);
+    /**
+     * @brief 	Define all numerical methods which are used in this case.
+     */
+    SimpleDynamics<NormalDirectionFromBodyShape> wall_boundary_normal_direction(wall_boundary);
+    /** delete outflow particles */
+    BodyAlignedBoxByCell left_disposer(
+        water_block, makeShared<AlignedBoxShape>(Transform(Rotation2d(Pi), Vec2d(left_bidirectional_translation)), bidirectional_buffer_halfsize));
+    SimpleDynamics<fluid_dynamics::DisposerOutflowDeletion> left_disposer_outflow_deletion(left_disposer, xAxis);
+    BodyAlignedBoxByCell right_disposer(
+        water_block, makeShared<AlignedBoxShape>(Transform(Vec2d(right_bidirectional_translation)), bidirectional_buffer_halfsize));
+    SimpleDynamics<fluid_dynamics::DisposerOutflowDeletion> right_disposer_outflow_deletion(right_disposer, xAxis);
+    /** surface particle identification */
+    InteractionWithUpdate<SpatialTemporalFreeSurfaceIndicationComplex>
+        free_stream_surface_indicator(water_block_inner, water_block_contact);
```

Comment:
should it be called free_stream_surface_indicator here?

Reply:
yes, this name is not strictly accurate,  it should be called boundary indicator, and i will rename it